src/v/cloud_storage/manifest.h

```diff
@@ -0,0 +1,202 @@
+
+struct serialized_json_stream {
+    ss::input_stream<char> stream;
+    size_t size_bytes;
+};
+
+enum class manifest_type {
+    topic,
+    partition,
+};
+
+class base_manifest {
+public:
+    /// Update manifest file from input_stream (remote set)
+    virtual ss::future<> update(ss::input_stream<char>&& is) = 0;
+
+    /// Serialize manifest object
+    ///
+    /// \return asynchronous input_stream with the serialized json
+    virtual serialized_json_stream serialize() const = 0;
+
+    /// Manifest object name in S3
+    virtual remote_manifest_path get_manifest_path() const = 0;
+
+    /// Get manifest type
+    virtual manifest_type get_manifest_type() const = 0;
+
+    /// Compare two manifests for equality
+    bool operator==(const base_manifest& other) const = default;
+};
+
+/// Manifest file stored in S3
+class manifest final : public base_manifest {
+public:
+    struct segment_meta {
+        bool is_compacted;
+        size_t size_bytes;
+        model::offset base_offset;
+        model::offset committed_offset;
+
+        auto operator<=>(const segment_meta&) const = default;
+    };
+    using key = segment_name;
+    using value = segment_meta;
+    using segment_map = absl::btree_map<key, value>;
+    using const_iterator = segment_map::const_iterator;
+
+    /// Create empty manifest that supposed to be updated later
+    manifest();
+
+    /// Create manifest for specific ntp
+    explicit manifest(model::ntp ntp, model::revision_id rev);
+
+    /// Manifest object name in S3
+    remote_manifest_path get_manifest_path() const override;
+
+    /// Segment file name in S3
+    remote_segment_path get_remote_segment_path(const segment_name& name) const;
+
+    /// Get NTP
+    const model::ntp& get_ntp() const;
+
+    // Get last offset
+    const model::offset get_last_offset() const;
+
+    /// Get revision
+    model::revision_id get_revision_id() const;
+
+    /// Return iterator to the begining(end) of the segments list
+    const_iterator begin() const;
+    const_iterator end() const;
+    size_t size() const;
+
+    /// Check if the manifest contains particular segment
+    bool contains(const segment_name& obj) const;
+
+    /// Add new segment to the manifest
+    bool add(const segment_name& key, const segment_meta& meta);
+
+    /// Get segment if available or nullopt
+    const segment_meta* get(const segment_name& key) const;
+
+    /// Get insert iterator for segments set
+    std::insert_iterator<segment_map> get_insert_iterator();
+
+    /// Return new manifest that contains only those segments that present
+    /// in local manifest and not found in 'remote_set'.
+    ///
+    /// \param remote_set the manifest to compare to
+    /// \return manifest with segments that doesn't present in 'remote_set'
+    manifest difference(const manifest& remote_set) const;
+
+    /// Update manifest file from input_stream (remote set)
+    ss::future<> update(ss::input_stream<char>&& is) override;
```

Comment:
if you take it by vaule, the compiler will complain if youdon't move it. but this is ok too. though this is tricky w/ ref input streams.